src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Data
+{
+    /// <summary>
+    /// Represents the <a href="https://en.wikipedia.org/wiki/Confusion_matrix">confusion matrix</a> of the classification results.
+    /// </summary>
+    public sealed class ConfusionMatrix
+    {
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Precision">precision</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassPrecision { get; }
+
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Recall">recall</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassRecall { get; }
+
+        /// <summary>
+        /// The confsion matrix counts for the combinations actual class/predicted class.
+        /// The actual classes are in the rows of the table, and the predicted classes in the columns.
+        /// </summary>
+        public ImmutableArray<ImmutableArray<double>> ConfusionTableCounts { get; }
+
+        /// <summary>
+        /// The indicators of the predicted classes.
+        /// It might be the classes names, or just indices of the predicted classes, if the name mapping is missing.
+        /// </summary>
+        public int NumberOfPredictedClasses { get; }
+
+        /// <summary>
+        /// The <see cref="DataViewSchema.Annotations"/> associated with the Confusion Matrix counts.
+        /// It contains information about the predicted classes, if that information is available.
+        /// It can be the classes names, or their indices.
+        /// </summary>
+        public DataViewSchema.Annotations ClassIndicators { get; }
```

Comment:
> public DataViewSchema.Annotations ClassIndicators { get; } [](start = 7, length = 59)

@tfinley@gmail.com is exposing this desirable?

Reply:
I'd prefer to not? AFAIK you get the class indicators out of the column they pass to the evaluate method. So why store it here? Why not just say, "get the indicators?" What does this API buy you? So my initial reaction is, "no, why is this object so complicated?"

---
In reply to: [275067993](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275067993) [](ancestors = 275067993)

Reply:
I mean, if I think about a confusion matrix, all I really want is the matrix, maybe a few summary statistics. Making this API for this limited object responsible for querying the metadata on the predicted label column *when we already have APIs to do that that are more general* seems like it is just making the story harder than it needs to be.

---
In reply to: [275473328](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275473328) [](ancestors = 275473328,275067993)